clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -168,76 +329,686 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
+        if (nameFromMetadataCache != null) {
+            // Add topic name to local cache, so it can be reused if included in a next target
+            // assignment if metadata cache not available.
+            assignedTopicNamesCache.put(topicId, nameFromMetadataCache);
+            return Optional.of(nameFromMetadataCache);
+        } else {
+            // Topic ID was not found in metadata. Check if the topic name is in the local
+            // cache of topics currently assigned. This will avoid a metadata request in the
+            // case where the metadata cache may have been flushed right before the
+            // revocation of a previously assigned topic.
+            String nameFromSubscriptionCache = assignedTopicNamesCache.getOrDefault(topicId, null);
+            return Optional.ofNullable(nameFromSubscriptionCache);
+        }
+    }
+
+    /**
+     * Build set of TopicPartition for the partitions included in the heartbeat topicPartitions,
+     * and using the given topic name.
+     */
+    private SortedSet<TopicIdPartition> buildAssignedPartitionsWithTopicName(
+            Uuid topicId,
+            String topicName,
+            List<Integer> topicPartitions) {
+        SortedSet<TopicIdPartition> assignedPartitions =
+                new TreeSet<>(TOPIC_ID_PARTITION_COMPARATOR);
+        topicPartitions.forEach(tp -> {
+            TopicIdPartition topicIdPartition = new TopicIdPartition(
+                    topicId,
+                    new TopicPartition(topicName, tp));
+            assignedPartitions.add(topicIdPartition);
+        });
+        return assignedPartitions;
+    }
+
+    /**
+     * Revoke partitions. This will:
+     * <ul>
+     *     <li>Trigger an async commit offsets request if auto-commit enabled.</li>
+     *     <li>Invoke the onPartitionsRevoked callback if the user has registered it.</li>
+     * </ul>
+     *
+     * This will wait on the commit request to finish before invoking the callback. If the commit
+     * request fails, this will proceed to invoke the user callbacks anyway,
+     * returning a future that will complete or fail depending on the callback execution only.
+     *
+     * @param revokedPartitions Partitions to revoke.
+     * @return Future that will complete when the commit request and user callback completes.
+     */
+    private CompletableFuture<Void> revokePartitions(Set<TopicPartition> revokedPartitions) {
+        log.info("Revoking previously assigned partitions {}", Utils.join(revokedPartitions, ", "));
+
+        logPausedPartitionsBeingRevoked(revokedPartitions);
+
+        // Mark partitions as pending revocation to stop fetching from the partitions (no new
+        // fetches sent out, and no in-flight fetches responses processed).
+        markPendingRevocationToPauseFetching(revokedPartitions);
+
+        // Future that will complete when the revocation completes (including offset commit
+        // request and user callback execution)
```

Comment:
nit: I have noticed that most of the comments end with a period but not all of them. It may be good to be consistent.